picard/ui/itemviews/custom_columns/README_API.md

```diff
@@ -0,0 +1,247 @@
+- `TransformProvider(base: ColumnValueProvider, transform: Callable[[str], str])`: applies a string transform.
+- `CallableProvider(func: Callable[[Item], str])`: wraps a Python callable.
+- Script provider is created via `make_script_column(...)` (do not instantiate directly).
+
+Factory helpers return a `CustomColumn` and infer a sane `sort_type` when possible:
+- `make_field_column(...)`
+- `make_script_column(...)` (tunable: `max_runtime_ms`, `cache_size`, optional parser or factory)
+- `make_transformed_column(...)`
+- `make_callable_column(...)`
+- `make_provider_column(...)`
+
+`CustomColumn` signature:
+```python
+CustomColumn(title: str,
+             key: str,
+             provider: ColumnValueProvider,
+             width: int | None = None,
+             align: ColumnAlign = ColumnAlign.LEFT,
+             sort_type: ColumnSortType = ColumnSortType.TEXT,
+             always_visible: bool = False)
+```
+
+---
+
+### Persistence utilities
+
+Serialize specs to config and (optionally) auto-register columns.
+
+```python
+from picard.ui.itemviews.custom_columns.storage import (
+    CustomColumnSpec, CustomColumnKind, TransformName,
+    build_column_from_spec,
+    load_specs_from_config, save_specs_to_config,
+    add_or_update_spec, delete_spec_by_key, get_spec_by_key,
+    register_and_persist, unregister_and_delete,
+    load_persisted_columns_once,
+)
+
+# Create and persist a script spec
+spec = CustomColumnSpec(
+    title="Artist – Title",
+    key="artist_title_script",
+    kind=CustomColumnKind.SCRIPT,
+    expression=script,
+    width=280,
+    align="LEFT",
+    add_to_file_view=False,
+    add_to_album_view=True,
+    insert_after_key="title",
+)
+register_and_persist(spec)  # saves to config and registers in views
+
+# Load and register all saved specs once (idempotent)
+load_persisted_columns_once()
+
+# Remove and delete
+unregister_and_delete("artist_title_script")
+```
+
+Notes:
+- `CustomColumnSpec.align` accepts "LEFT" or "RIGHT" (mapped to `ColumnAlign`).
+- `CustomColumnSpec.kind`: `FIELD`, `SCRIPT`, or `TRANSFORM`.
+- `TRANSFORM` specs use `expression` as the base field and optional `transform: TransformName`.
+- Registry insertion uses the spec’s `add_to_file_view`, `add_to_album_view`, and `insert_after_key`.
+
+---
+
+### Field keys and scripting
+
+- Field keys are the same strings used with `obj.column(key)` and Picard variables without percent signs (e.g. `title`, `albumartist`, `~bitrate`).
+- Script expressions use the standard Picard scripting language (e.g. `$if()`, `$if2()`, `%artist%`).
+- See `picard.const.tags.ALL_TAGS` for the authoritative list of variables.
+
+---
+
+### Runtime & safety
+
+- Script provider has configurable `max_runtime_ms` and internal caching; errors return empty strings rather than raising.
+- `registry.register` is UI-safe after the main window has initialized; re-entrant calls replace existing keys.
+- `registry.unregister(key)` removes from both views (if present).
```

Comment:
Keeping this in sync with actual code will be a pain to maintain (due to code examples mainly).

Reply:
Ideally it should be generated from code instead, so it can easily be synced.